Ai player

ksabbak opened the pull request.

ksabbak committed:
Add make-move function that returns a random open space

This involved adding a open-spaces function to the board ns, which allowed for a refactor of full? given the logic repeated.

ksabbak committed:
Refactor

ksabbak committed:
Fix merge conflict

ksabbak committed:
Integrate vs. computer

Players now play against a very stupid ai.

hnlee commented on src/clojure_tictactoe/game/game_play.clj:
```diff
@@ -1,34 +1,32 @@
+              (recur new-board))
+        (if-let [results (rules/assess-win board)]
+                (println (end-printer/game-won-message results))
+                (println end-printer/game-tie-message))))))
```
`game-play` should eventually not need to know that the UI is a console (i.e. not call `println` which is console-dependent directly).

ksabbak replied:
I agree, but I have some questions surrounding this. Specifically, do you have any suggestions for how to accomplish this without just wrapping `println`?

hnlee commented on src/clojure_tictactoe/game/game_rules.clj:
```diff
@@ -27,8 +26,7 @@
-(defn game-over?
-  [board]
+(defn game-over? [board]
   (not (not (or
```
I know you talked with Byron about this on the last PR, but I'm still not entirely clear on why this can't be simplified. Can `board/board-full? board` or `assess-win board` be made to always return booleans?

ksabbak replied:
I reworked `assess-win` in order to address this. It no longer exists and has been replaced with `assess-winner` and `winner`. `board-full?` always returned booleans, but both of them had to always return booleans in order to ensure the not-not was unnecessary.

hnlee commented on src/clojure_tictactoe/game/game_play.clj:
```diff
@@ -1,34 +1,32 @@
+  (instructions-printer/print-game-intro)
+  (input-getter/continue-to-game)
+  (let [board (board/render-empty-board 9)]
+    (board-printer/print-board board)
```
What you're doing above here is not strictly `game-loop`. Maybe extract them out into an `initialize-game` function that returns a `board` you can pass to `game-loop`.

ksabbak replied:
👍

hnlee commented on src/clojure_tictactoe/game/players.clj:
```diff
@@ -9,3 +13,17 @@
+    (let [move (input-getter/get-player-choice)]
+      (if (board/space-is-open? move board)
+          move
+          (do (println "Sorry, that looks taken, try again")
```
This part looks like it should be moved to a namespace that knows about UI.

ksabbak replied:
Very true. I moved it, but I think I need help figuring out how to untangle board, players and input-getter.

hnlee reviewed (changes requested):
Overall, looks much cleaner. I think the parts that you are unhappy with are the parts where you need to remind yourself of SRP. What is this function responsible for? What does it need to know about? Answering those questions may help untangle things.

ksabbak committed:
Refactor to address comments on PR 13

https://github.com/ksabbak/ClojureTTT/pull/13

Game Rules's intense overhaul was partly in response to a future feature of 4x4 board. While separating assess-win and winner? could have been done independent of the other refactor, it would have led to more work down the road in undoing and redoing certain things.
